카드에 멤버 추가/삭제 API 구현

Park-SooYeon opened the pull request:
# 카드에 멤버 추가/삭제 API 구현

## 📎 해당 이슈

이슈 링크 (#188)

## 🛠 구현 내용 

- `PUT /api/card/{cardId}/member` API 구현
- 요청 성공 시, 204 응답 반환
- 보드에 속하지 않은 사용자가 요청 시, 403 응답 반환
- 멤버를 추가할 카드가 존재하지 않을 경우, 404 응답 반환
- 추가할 멤버가 보드에 속한 사용자가 아닐 경우, 409 응답 반환

## 🙋‍ 리뷰어 참고 사항

Park-SooYeon committed:
[FEAT] 카드에 멤버 추가 API 서비스 함수 추가

- PUT /api/card/{cardId}/member 요청을 처리 할 addMemberToCardByUserIds() 서비스 함수 추가
- memberRepository 추가
- 추가하려는 멤버가 보드에 속한 멤버들인지 확인하는 existUsersByBoardId() 함수 추가

Park-SooYeon committed:
[TEST] addMemberToCardByUserIds 함수 테스트 코드 추가

Park-SooYeon committed:
[FEAT] PUT /api/card/{cardId}/member 라우터 추가

- 카드에 멤버 추가 API 요청을 위한 라우터 추가
- validation에 사용할 MemberDto 추가

Park-SooYeon committed:
[TEST] PUT /api/card/{cardId}/member 테스트 코드 추가

- 요청 성공 시, 204 응답 반환
- 보드에 속하지 않은 사용자가 요청 시, 403 응답 반환
- 멤버를 추가할 카드가 존재하지 않을 경우, 404 응답 반환
- 추가할 멤버가 보드에 속한 사용자가 아닐 경우, 409 응답 반환

youngxpepp commented on server/src/service/CardService.js:
```diff
@@ -131,4 +131,36 @@ export class CardService extends BaseService {
+            return;
+        }
+
+        if (!isExistUser) throw new ForbiddenError('no access to this card');
```
```suggestion
        if (!isExistUser) throw new ForbiddenError('no access to this card');
        
        if (userIds.length === 0) {
            await this.memberRepository.delete({ card: cardId });
            return;
        }
```
두 개의 로직이 순서가 바뀌어야 된다고 생각합니당. 권한 검사를 하고 나서 수정해야 할듯 해요

Park-SooYeon replied:
commit 770c177
수정했습니다!

youngxpepp commented on server/src/dao/CustomBoardRepository.js:
```diff
@@ -49,6 +49,22 @@ export class CustomBoardRepository extends BaseRepository {
         return isAuth.length !== 0;
     }
 
+    async existUsersByBoardId({ boardId, userIds }) {
+        const creatorCount = await this.createQueryBuilder('board')
+            .where('board.id=:boardId', { boardId })
+            .andWhere('board.creator_id IN(:...userIds)', { userIds })
+            .getCount();
+        const isExistinvitation = await this.createQueryBuilder('board')
+            .innerJoinAndSelect('board.invitations', 'invitation')
+            .where('board.id=:boardId', { boardId })
+            .andWhere('invitation.user_id IN(:...userIds)', { userIds })
+            .getOne();
+
+        const invitationCount = isExistinvitation ? isExistinvitation.invitations.length : 0;
+
+        return userIds.length === creatorCount + invitationCount;
+    }
```
어떤 역할을 하는 메소드인지 궁금해요. 설명 부탁 드려도 될까요?🙏

Park-SooYeon replied:
카드 멤버에 추가하려는 멤버 중, 한명이라도 보드에 속한 사용자가 아니라면 추가를 할 수 없을 것이라고 생각해서 멤버 중, 보드에 속하지 않은 사용자가 있는지 확인하는데 사용하려고 만든 함수입니다!

youngxpepp replied:
말을 듣고 코드를 보니 이해가 되네요! 감사합니다.

youngxpepp commented on server/src/dto/MemberDto.js:
```diff
@@ -0,0 +1,13 @@
+
+    user;
+
+    @ValidateIf((o) => o.user)
+    @IsArray()
+    userIds;
```
해당 부분은 멤버 엔티티와 관련이 없는 것처럼 보이는데, 넣은 의도가 있을까요?

Park-SooYeon replied:
카드 멤버 추가 body의 userIds값을 체크할 때 사용하려고 추가했습니다.
DTO를 이렇게 정의하고 validator 함수를 호출할 때는 아래와 같이 해서 userIds 필드의 validation을 체크하고 싶으면 user에 true를 넣도록하게했어요!
``` javascript
validator(MemberDto, { user: true, userIds });
```
이렇게하면 필요할 때마다 validate하는 조건을 추가할 수 있지 않을까해서 이렇게 사용해봤습니다!
ValidateIf를 좀 다른 방법으로 사용해본 것인데..아직 엔티티를 따르는 DTO를 어떻게 validate하는데 사용해야할지 잘 모르겠네요..그냥 groups를 사용할걸 그랬나봐요.🤔

youngxpepp replied:
![image](https://user-images.githubusercontent.com/47842964/101581823-2474db00-3a1e-11eb-8bbf-495f26c0a8e2.png)
멘토님의 조언대로라면, DTO는 엔티티에 기반해서 만드는 것이고 복합적인 요청 필드는 다른 클래스나 인터페이스를 통해서 관리되는 듯 하네요. 수연님 상황이라면 userIds를 분리하는게 맞다고 생각합니다.

Park-SooYeon replied:
그렇다면 MemberDto를 상속받은 클래스를 따로 만들어서 validation을 하는게 맞을것 같네요.

youngxpepp replied:
멘토님께서 말씀해주신 각각의 항목에 따라 DTO에 따른 validation을 인용해서 요청을 만들어 보자면, 아래와 같은 방식으로 해야 하지 않을까요? (예시에요!)

```javascript
class AddMemberRequestBody {
    @IsArray()
    userIds;

    // 재귀적으로 검증하는 데코레이터가 있을 것 같아요...
    userDto;

    // 재귀적으로 검증하는 데코레이터가 있을 것 같아요...
    memberDto;
}
```

Park-SooYeon replied:
인터페이스를 활용하고 받은후에 validation을 한다는 것을 저는 엔티티에 종속적인 DTO를 상속을 받는 필요한 데이터만 추가하는 것으로 이해했습니다.
건홍님 방식대로 클래스를 만들게 되면 하위에 추가한 DTO의 모든 변수들이 body나 param에 있어야만 validation을 통과할 수 있지 않나요?
지금 API를 예를들면 cardId와 userIds가 필수값이고 member 테이블의 id와 user_id값은 없는데 memberDto를 필드로 추가하고 memberDto 전체에 따른 validation을 수행하게 되면 id, user_id 값이 존재하지 않으니 무조건 validation이 실패할 것 같다는 생각이 드네요.🤔
건홍님 생각은 어떠신가요??
일단 userIds를 memberDto에 둔 것은 분리해야할 것이라 생각하지만..
어떻게 해야 엔티티에 종속적인 DTO로 각 필드에 따른 validation을 선택적으로 해줄 수 있을지 잘 모르겠네요😵

youngxpepp replied:
제 방법을 쓰면 body, param에 종속이 되어야 하는건 맞겠네요. id, user_id 값이 존재하지 않는 건 groups를 통해 해결할 수 있을 거라고 생각했습니다!
만약에 CardDto도 필요하고 BoardDto도 필요한 상황이라면, 두 개의 DTO를 상속해야 하는 상황인데 이런 경우는 어떻게 해야 할까요? JS에서 다중 상속을 지원할지 안 할지 모르겠네요.

Park-SooYeon replied:
java의 경우에도 인터페이스가 아닌 클래스는 다중 상속이 불가능하지 않나요?
js는 인터페이스를 지원하지 않으니 상속받아 사용하기에는 확실히 한계가 있긴하겠네요..
엔티티에 종속된 DTO 활용이란..너무 어렵네요..😂
request body랑 param에 종속되는 dto를 계속 생성하고 있는 느낌입니다..
일단은 새로 멤버 추가시에 사용할 body dto 만들고 계속 진행해보겠습니다..😢

youngxpepp replied:
넵 java도 다중 상속이 불가능해요ㅠㅠ
더군다나 인터페이스라고 해도 문제가 되는게, class-validator의 데코레이터가 TS의 인터페이스에 적용이 될지도 모르는거구...
cardId와 userIds 같은 경우엔 엔티티 기반 dto를 사용하지 않고 검증을 해도 무방하다고 생각해요. 빠르게 개발해야 하는거니까 dto는 좀 더 생각해봅시다

Park-SooYeon replied:
commit 1346321
dto 분리했습니다!

youngxpepp reviewed (changes requested):
PUT으로 구현하니까, 로직이 너무 복잡해지네요😓
어찌해야 할까요? 너무 고생하시네요..

Park-SooYeon committed:
[FIX] 리뷰에 따른 코드 변경

- 리뷰에 따라 권한 검사 순서 변경

Park-SooYeon committed:
[FIX] develop/server pull conflict 해결

- pull conflict 해결

Park-SooYeon committed:
[FIX] 리뷰에 따른 DTO 추가 및 변경

- userIds 필드 분리를 위해 AddMemberBodyDto 추가
- dto 추가에 따라 cardId도 validation 체크하도록 추가

youngxpepp reviewed (approved):
고생하셨습니다!
DTO 관련된건 멘토님께 질문해봐요

dh00n commented on server/src/dao/CustomBoardRepository.js:
```diff
@@ -49,6 +49,22 @@ export class CustomBoardRepository extends BaseRepository {
+
+        const invitationCount = isExistinvitation ? isExistinvitation.invitations.length : 0;
+
+        return userIds.length === creatorCount + invitationCount;
```
기발하네효 👍👍

dh00n reviewed (approved):
💯💯
